File: base64_url.php
<?php
// base64_url.php — PHP версия

function copyToClipboard($text) {
    if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
        $cmd = 'echo ' . escapeshellarg($text) . ' | clip';
        exec($cmd, $output, $return);
        return $return === 0;
    } elseif (PHP_OS === 'Darwin') {
        $cmd = 'echo ' . escapeshellarg($text) . ' | pbcopy';
        exec($cmd, $output, $return);
        return $return === 0;
    } else {
        $cmd = 'echo ' . escapeshellarg($text) . ' | xclip -selection clipboard';
        exec($cmd, $output, $return);
        if ($return === 0) return true;
        $cmd = 'echo ' . escapeshellarg($text) . ' | xsel -b';
        exec($cmd, $output, $return);
        return $return === 0;
    }
}

function urlSafeBase64Encode($data) {
    return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
}

function urlSafeBase64Decode($data) {
    $data = strtr($data, '-_', '+/');
    $padding = 4 - (strlen($data) % 4);
    if ($padding != 4) $data .= str_repeat('=', $padding);
    return base64_decode($data);
}

$input = '';
$filePath = '';
$decode = false;
$output = '';
$noCopy = false;

$args = array_slice($argv, 1);
for ($i = 0; $i < count($args); $i++) {
    if ($args[$i] == '--input') {
        $input = $args[++$i];
    } elseif ($args[$i] == '--file') {
        $filePath = $args[++$i];
    } elseif ($args[$i] == '--decode') {
        $decode = true;
    } elseif ($args[$i] == '--output') {
        $output = $args[++$i];
    } elseif ($args[$i] == '--no-copy') {
        $noCopy = true;
    } elseif (!str_starts_with($args[$i], '-')) {
        if (empty($input) && empty($filePath)) {
            $input = $args[$i];
        }
    }
}

echo "\033[36m🔐 Base64 URL-safe Encoder (PHP)\033[0m\n";

$result = '';

if (!empty($filePath)) {
    echo "📂 Обработка файла: $filePath\n";
    $content = file_get_contents($filePath);
    if ($content === false) {
        echo "\033[31m❌ Ошибка чтения файла.\033[0m\n";
        exit(1);
    }
    if ($decode) {
        $result = urlSafeBase64Decode($content);
    } else {
        $result = urlSafeBase64Encode($content);
    }
} elseif (!empty($input)) {
    echo "📝 Входные данные: $input\n";
    if ($decode) {
        $result = urlSafeBase64Decode($input);
    } else {
        $result = urlSafeBase64Encode($input);
    }
} else {
    echo "📝 Чтение из STDIN (Ctrl+D для окончания)\n";
    $data = file_get_contents('php://stdin');
    if ($data === false || $data === '') {
        echo "\033[33m⚠️ Пустой ввод.\033[0m\n";
        exit(1);
    }
    if ($decode) {
        $result = urlSafeBase64Decode($data);
    } else {
        $result = urlSafeBase64Encode($data);
    }
}

echo "\033[32mРезультат:\033[0m\n";
echo $result . "\n";
echo "\n";

if (!$noCopy) {
    if (copyToClipboard($result)) {
        echo "\033[32m✅ Результат скопирован в буфер обмена!\033[0m\n";
    } else {
        echo "\033[33m⚠️ Не удалось скопировать в буфер обмена.\033[0m\n";
    }
}

if (!empty($output)) {
    file_put_contents($output, $result);
    echo "\033[32m💾 Сохранено в $output\033[0m\n";
}
?>

File: base64_url.go
// base64_url.go — Go версия

package main

import (
	"bufio"
	"encoding/base64"
	"flag"
	"fmt"
	"io"
	"os"
	"os/exec"
	"runtime"
	"strings"
)

func copyToClipboard(text string) bool {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "windows":
		cmd = exec.Command("clip")
	case "darwin":
		cmd = exec.Command("pbcopy")
	case "linux":
		if _, err := exec.LookPath("xclip"); err == nil {
			cmd = exec.Command("xclip", "-selection", "clipboard")
		} else if _, err := exec.LookPath("xsel"); err == nil {
			cmd = exec.Command("xsel", "-b")
		} else {
			return false
		}
	default:
		return false
	}
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return false
	}
	if err := cmd.Start(); err != nil {
		return false
	}
	stdin.Write([]byte(text))
	stdin.Close()
	return cmd.Wait() == nil
}

func encodeString(data string) string {
	return base64.URLEncoding.EncodeToString([]byte(data))
}

func encodeFile(filename string) (string, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return base64.URLEncoding.EncodeToString(content), nil
}

func decodeString(data string) (string, error) {
	// URLEncoding автоматически игнорирует padding
	decoded, err := base64.URLEncoding.DecodeString(data)
	if err != nil {
		return "", err
	}
	return string(decoded), nil
}

func main() {
	var input string
	var filePath string
	var decode bool
	var output string
	var noCopy bool

	flag.StringVar(&input, "input", "", "Входная строка")
	flag.StringVar(&filePath, "file", "", "Файл для обработки")
	flag.BoolVar(&decode, "decode", false, "Декодировать")
	flag.StringVar(&output, "output", "", "Сохранить результат в файл")
	flag.BoolVar(&noCopy, "no-copy", false, "Не копировать в буфер обмена")
	flag.Parse()

	fmt.Println("\x1b[36m🔐 Base64 URL-safe Encoder (Go)\x1b[0m")

	var result string
	var err error

	if filePath != "" {
		fmt.Printf("📂 Обработка файла: %s\n", filePath)
		if decode {
			content, err := os.ReadFile(filePath)
			if err != nil {
				fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
				os.Exit(1)
			}
			result, err = decodeString(strings.TrimSpace(string(content)))
		} else {
			result, err = encodeFile(filePath)
		}
	} else if input != "" {
		fmt.Printf("📝 Входные данные: %s\n", input)
		if decode {
			result, err = decodeString(input)
		} else {
			result = encodeString(input)
		}
	} else {
		fmt.Println("📝 Чтение из STDIN (Ctrl+D для окончания)")
		reader := bufio.NewReader(os.Stdin)
		content, err := io.ReadAll(reader)
		if err != nil {
			fmt.Printf("\x1b[31m❌ Ошибка чтения: %v\x1b[0m\n", err)
			os.Exit(1)
		}
		if len(content) == 0 {
			fmt.Println("\x1b[33m⚠️ Пустой ввод.\x1b[0m")
			os.Exit(1)
		}
		if decode {
			result, err = decodeString(strings.TrimSpace(string(content)))
		} else {
			result = encodeString(string(content))
		}
	}

	if err != nil {
		fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
		os.Exit(1)
	}

	fmt.Printf("\x1b[32mРезультат:\x1b[0m\n")
	fmt.Println(result)
	fmt.Println()

	if !noCopy {
		if copyToClipboard(result) {
			fmt.Println("\x1b[32m✅ Результат скопирован в буфер обмена!\x1b[0m")
		} else {
			fmt.Println("\x1b[33m⚠️ Не удалось скопировать в буфер обмена.\x1b[0m")
		}
	}

	if output != "" {
		err = os.WriteFile(output, []byte(result), 0644)
		if err != nil {
			fmt.Printf("\x1b[31m❌ Ошибка сохранения: %v\x1b[0m\n", err)
		} else {
			fmt.Printf("\x1b[32m💾 Сохранено в %s\x1b[0m\n", output)
		}
	}
}
